apis/Google.Storage.V1/Google.Storage.V1/SignedUrlUtils.cs

```diff
@@ -0,0 +1,396 @@
+            }
+            var extensionHeaders = GatherExtensionHeaders(requestHeaders, contentHeaders);
+
+            var contentMD5 = GetFirstHeaderValue(contentHeaders, "Content-MD5");
+            var contentType = GetFirstHeaderValue(contentHeaders, "Content-Type");
+
+            var signatureLines = new List<string>
+            {
+                (requestMethod ?? HttpMethod.Get).ToString(),
+                contentMD5,
+                contentType,
+                timestamp
+            };
+            signatureLines.AddRange(extensionHeaders.Select(
+                header => $"{header.Key}:{string.Join(", ", header.Value)}"));
+            signatureLines.Add(resourcePath);
+
+            var signature = CreateSignature(string.Join("\n", signatureLines), credentials);
+
+            var queryParameters = new List<string> { $"GoogleAccessId={credentials.Id}" };
+            if (timestamp != null)
+            {
+                queryParameters.Add($"Expires={timestamp}");
+            }
+            queryParameters.Add($"Signature={signature}");
+            return $"{StorageHost}{resourcePath}?{string.Join("&", queryParameters)}";
+        }
+        
+        private static string Base64Encode(this byte[] value)
+        {
+            return Convert.ToBase64String(value);
+        }
+
+        private static string CreateSignature(string data, ServiceAccountCredential.Initializer initializer)
+        {
+            // TODO: This is taken from ServiceAccountCredential. Expose this somehow so we don't need to duplicate logic.
+            var hashAlg = SHA256.Create();
+            byte[] assertionHash = hashAlg.ComputeHash(Encoding.UTF8.GetBytes(data));
+#if NETSTANDARD1_3
+            var sigBytes = initializer.Key.SignHash(assertionHash, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
+#else
+            const string Sha256Oig = "2.16.840.1.101.3.4.2.1";
+            var sigBytes = initializer.Key.SignHash(assertionHash, Sha256Oig);
+#endif
+            return WebUtility.UrlEncode(Convert.ToBase64String(sigBytes));
+        }
+
+        private static SortedDictionary<string, StringBuilder> GatherExtensionHeaders(
+            Dictionary<string, IEnumerable<string>> requestHeaders,
+            Dictionary<string, IEnumerable<string>> contentHeaders)
+        {
+            // These docs indicate how to include extension headers in the signature, but they're not exactly
+            // correct (values must be trimmed, newlines are replaced with empty strings, not whitespace, and
+            // values are concatenated with ", " instead of ",", but not when joining request and content headers).
+            // https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-extension-headers
+            var extensionHeaders = new SortedDictionary<string, StringBuilder>();
+
+            if (requestHeaders != null)
+            {
+                GatherExtensionHeaders(requestHeaders, extensionHeaders);
```

Comment:
I would change the name of one of the methods to make it clear that this isn't a recursive call.